### PR TITLE
Updating SQLAlchemy 1.4 to 2.0

### LINKED ISSUE #581: API: Migrate SQLAlchemy 1.4.x to 2.0.x
### Overview
SQLAlchemy is part of the data persistence layer of the API. The API is currently using a version from the 1.4.x series. Migrate the SQLAlchemy from 1.4.x to 2.0.x.

### Action Items

- [ ] Review the [SQLAlchemy 1.4 to 2.0 migration guide](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html) 
- [ ] Migrate the API code to use SQLAlchemy 2.0
- [ ] Help merge existing API PRs with new changes

### Resources/Instructions
Use the migration guide to help perform the migration: https://docs.sqlalchemy.org/en/20/changelog/migration_20.html

### PR DESCRIPTION
**Pull request by shawnkoong**
Closes #581 

### What changes did you make?
  - Removed SQLAlchemy from `pyproject.toml` to make it a transitive dependency of alembic
  - Converted SQLAlchemy 1.4 syntax to 2.0 syntax (mainly `session.query` to `session.execute` followed by `select`)
### Rationale behind the changes?
  - With 2.0 comes new deprecations; functions changing location/changing import paths
  - ORM related code will still be functional with 1.x style, but will not be supported or included in future guide/tutorial.
### Testing done for these changes
  - test cases and app ran and checked with deprecation warning flag on
  - currently, the test cases only included those for `service_provider_repository`
### What did you learn or can share that is new?(optional)
  - the migration guide can be found [here](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html)
  - also included a link to a table showing the translation from 1.x to 2.0 in #581

### CODE REVIEW

File: api/openapi_server/openapi/paths/users.yaml
Comment (shawnkoong) [inline api/openapi_server/openapi/paths/users.yaml:1]: had to change `operationId` here from `delete` to `delete_user` to match with the method name in `api/openapi_server/controllers/users_controller.py` for name collision. We if do separate controller/repository, the method name can be reverted as with `operationId`

Comment (paulespinosa) [reply api/openapi_server/openapi/paths/users.yaml:1]: The new name is a good idea. The `delete_user` function looks more descriptive.

File: api/openapi_server/repositories/service_provider_repository.py [hunk 1]
```diff
@@ -54,14 +56,14 @@ def get_service_provider_by_id(self, provider_id: int) -> Optional[HousingProgra
...
-            return session.query(HousingProgramServiceProvider).get(provider_id)
+            return session.get(HousingProgramServiceProvider, provider_id)
             
     def get_service_providers(self) -> List[HousingProgramServiceProvider]:
         """
         Get a list of all housing program service providers.
         """
         with DataAccessLayer.session() as session:
-            return session.query(HousingProgramServiceProvider).all()
+            return session.execute(select(HousingProgramServiceProvider)).scalars().all()
```
Comment (paulespinosa) [inline api/openapi_server/repositories/service_provider_repository.py:66]: Why does SQLA have two ways of doing this `session.scalars(...).all()` and `session.execute(...).scalars().all()`

**Review by paulespinosa — COMMENTED** [review]
Thank you @shawnkoong! It looks good. Just had some questions.

File: api/openapi_server/controllers/auth_controller.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 from openapi_server.models.database import DataAccessLayer, User
 from sqlalchemy.orm import Session
 from sqlalchemy.exc import IntegrityError
+from sqlalchemy.future import select
```
[the remark this answers is not in the document]
Comment (shawnkoong) [reply api/openapi_server/controllers/auth_controller.py:15]: You're right, no need to use `future` in 2.0

File: api/openapi_server/repositories/service_provider_repository.py (hunk 1 above)
Comment (shawnkoong) [reply api/openapi_server/repositories/service_provider_repository.py:66]: The `session.execute()` is the more general, SQL like approach. This returns [Row](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row) objects (containing individual columns), which need to be unwrapped into the ORM objects (using `scalars()`).
`session.scalars()` skips the Row objects part and returns the ORM object directly.
Both are valid and equivalent, I just chose the one that was listed first in the translation table.

Is there a particular preference for one over another?

Comment (shawnkoong) [reply api/openapi_server/repositories/service_provider_repository.py:66]: https://docs.sqlalchemy.org/en/20/orm/queryguide/select.html#orm-queryguide-select-orm-entities

This portion of the doc talks about them both

Comment (Joshua-Douglas) [reply api/openapi_server/repositories/service_provider_repository.py:67]: Yeah that portion of the docs says that `session.scalars(...)` is equivalent to `session.execute(...).scalars()`. It is just a shortcut if you need to call the `scalars()` method on the `Result` object returned by a session statement execution. 

With that in mind, you can consolidate those three lines of code into a single return statement:
```
return session.scalars(select(HousingProgramServiceProvider)).all()
```

**Review by Joshua-Douglas — CHANGES_REQUESTED** [review]
Hey @shawnkoong, 

Changes look great. Just had a minor comment for the changes to the backend. The new query pattern is interesting - thanks for doing the research!

Before we can submit, it looks like there was an issue with your last merge from main in c95bd4c5de3585ebe587053dd5ea3b547e78f72c. That commit shows several changes to the frontend app that don't appear to be on `main`. For example the `app/src/components/dashboard/DashboardTask.tsx` file was deleted, but [that file is still on main](https://github.com/hackforla/HomeUniteUs/blob/main/app/src/components/dashboard/DashboardTask.tsx). 

Did you have a merge conflict while performing the merge? It might be a good idea to revert that commit and try the merge again. Let me know if you'd like me to help out.

**Comment by shawnkoong** [comment]
whoops, my tired brain must have done its own thing and unstaged them, will revert and fix

**Comment by Joshua-Douglas** [comment]
No worries @shawnkoong. It looks like there are still a couple more merge errors. You added 'encoder.py' and reverted the update to the 'connexion' version update that was made in PR #601.

Can you compare your branch with that PR and fix the conflicts? Thanks!

**Comment by shawnkoong** [comment]
> Can you compare your branch with that PR and fix the conflicts? Thanks!

@Joshua-Douglas ended up just resetting to the commit pre-merge. Fixed now

**Comment by Joshua-Douglas** [comment]
> > Can you compare your branch with that PR and fix the conflicts? Thanks!
> 
> 
> 
> @Joshua-Douglas ended up just resetting to the commit pre-merge. Fixed now 

Great idea!

**Review by Joshua-Douglas — APPROVED** [review]
All of the code changes appear to be consistent with the SQLAlchemy 2.0 migration guide. So glad to be on the latest version. Approved!

**Pull request merged**